<a href="https://colab.research.google.com/github/rikithreddy/dsworkshop/blob/main/Data_Science_Workshop_feb_5_2022.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Welcome to google colab

Some interesting features:
* Comment and Share it really quick!
* Free resources!
* Mounting gdrive to access your files!
* Text and code in the same place!


---



# Basics of google colab

## Text and code cells?
**This is a text cell**

Hello whats up

```python
a = 1
print(1)
```

In [2]:
# This is a code cell
a = 1
print(a)

1


## Prints are not needed in most of the cases

In [8]:
print(1)

1


## Try it yourself
Write some basic python code and get a feel of your notebook.

In [9]:
# Write some python code here :)
a = 1
b = 1
a+b

2

In [11]:
b = 2
print(b)

2


## Pro Tip
Sticking to good practices, always run cells in order.

# Data & Python!
```
What is data?
Where is data?
Who is data?
```

* Spreadsheets
* API calls
* JSON files
* CSV files
* YAML files
* SQL database


## That's ok but how do I get this data to python?

Using **pandas**

![pandas](http://4.bp.blogspot.com/-JYiKvzQgg4g/UQFnLfJR6UI/AAAAAAAAEVU/EnPgdxrRHXs/s1600/Panda-info-and+images+07.jpg)

## Pandas

### Official definition
```
Pandas is a fast, powerful, flexible and easy to use open source data analysis and manipulation tool,
built on top of the Python programming language.
```

### Layman's definition
```
Python object which can perform everything your Excel spreadsheet can.
```

## Code examples

Let's load this data

https://raw.githubusercontent.com/PyDataPune/Code-and-Coffee-2/master/matches.csv

**Data Credits**: [Mayank Mishra](https://github.com/mayankskb) (*pydata:Pune*)

In [12]:
import pandas as pd

df = pd.read_csv("https://raw.githubusercontent.com/PyDataPune/Code-and-Coffee-2/master/matches.csv")
df

city        date  ... win_by_wickets                       winner
0         Mumbai  2010-04-24  ...            9.0  Royal Challengers Bangalore
1          Delhi  2010-03-29  ...            NaN             Delhi Daredevils
2         Mumbai  2008-05-16  ...            8.0               Mumbai Indians
3           Pune  2016-04-22  ...            NaN  Royal Challengers Bangalore
4         Mumbai  2010-04-22  ...            NaN          Chennai Super Kings
..           ...         ...  ...            ...                          ...
741    Hyderabad  2018-04-12  ...            1.0          Sunrisers Hyderabad
742  East London  2009-05-04  ...            NaN          Chennai Super Kings
743    Bangalore  2008-05-28  ...            9.0               Mumbai Indians
744          NaN  2014-04-28  ...            5.0              Kings XI Punjab
745    Bangalore  2010-04-17  ...            NaN               Mumbai Indians

[746 rows x 20 columns]

# Head and Tail functions

In [23]:
df.head(1)

city        date  ... win_by_wickets                       winner
0  Mumbai  2010-04-24  ...            9.0  Royal Challengers Bangalore

[1 rows x 20 columns]

In [26]:
df.tail(3)

city        date  ... win_by_wickets           winner
743  Bangalore  2008-05-28  ...            9.0   Mumbai Indians
744        NaN  2014-04-28  ...            5.0  Kings XI Punjab
745  Bangalore  2010-04-17  ...            NaN   Mumbai Indians

[3 rows x 20 columns]

### Do it yourself

Get first 0 rows

Get last 747 rows
Note the dataframe just has 746 rows

Try to get 3,4,5 rows using head and tail

## Good practice

Chaining!

In [36]:
tdf = df.head(100)
tdf = tdf.tail(30)
tdf = tdf.head(2)
tdf

city        date  ... win_by_wickets               winner
70   Ranchi  2014-05-02  ...            NaN  Chennai Super Kings
71  Chennai  2011-05-04  ...            8.0  Chennai Super Kings

[2 rows x 20 columns]

In [44]:
(
  df
    .head(100)
    .tail(50)
    .head(2)
)

city        date  ... win_by_wickets               winner
50     Bangalore  2016-05-29  ...            NaN  Sunrisers Hyderabad
51  Johannesburg  2009-05-10  ...            7.0     Delhi Daredevils

[2 rows x 20 columns]

## Basic exploration

In [ ]:
df.describe()

id  overs       season  win_by_runs  win_by_wickets
count  7.460000e+02  746.0   746.000000   336.000000      398.000000
mean   7.135520e+05   20.0  2013.369973    29.883929        6.251256
std    2.845419e+05    0.0     3.326906    27.276827        1.821281
min    3.359820e+05   20.0  2008.000000     1.000000        1.000000
25%    5.012092e+05   20.0  2011.000000    11.000000        5.000000
50%    5.980485e+05   20.0  2013.000000    22.000000        6.000000
75%    9.809845e+05   20.0  2016.000000    39.000000        8.000000
max    1.178425e+06   20.0  2019.000000   146.000000       10.000000

## Counting matches

In [46]:
df.result.unique()

array([nan, 'no result', 'tie'], dtype=object)

In [45]:
df.result.value_counts()

tie          8
no result    4
Name: result, dtype: int64

## Your turn

* What are the unique values of `toss_decision`

* What does `value_counts` of `toss_decision` look like?
* Do those values add up to `746`?
* What can you conclude from that?

* Get rows 5-100 using head and tail functions and perform a descirbe()

* How many matches have first innings batters won?


* Get rows 50-200 using `head` and `tail` functions and perform a `describe()`

* Find value counts of `result`?


## Selection operations

#### Lets go back to lists in python

In [47]:
l = [0,1,2,3,4,5,6,7,8,9,10]
l


[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

In [48]:
l[1:4]

[1, 2, 3]

In [49]:
l[:5]

[0, 1, 2, 3, 4]

In [50]:
l[6:]

[6, 7, 8, 9, 10]

#### Your turn
Try getting an output of `[7,8,9]` using both first and last index

Try getting the output `[8,9,10]` using only first index

Try getting the output `[1,2,3,4,5]` using only last index

Get all elements but use slicing

#### Back to pandas

Head and tail operations are good but can we have more control?

Yes. `iloc`

In [58]:
df.iloc[4:6, 2:9]

dl_applied eliminator      id  overs player_of_match result  season
4        NaN        NaN  419163     20    DE Bollinger    NaN    2010
5        NaN        NaN  419136     20    KP Pietersen    NaN    2010

In [59]:
df.iloc[[4, 6, 7],:]

city        date  ... win_by_wickets                 winner
4     Mumbai  2010-04-22  ...            NaN    Chennai Super Kings
6  Bangalore  2008-05-19  ...            5.0       Delhi Daredevils
7    Kolkata  2016-05-04  ...            NaN  Kolkata Knight Riders

[3 rows x 20 columns]

* Get rows 100 to 200 with columns 1 to 5

* Get all rows with only columns 2,5

* Get last 5 columns with 8,100,201 rows

12

* Get rows 5-100 using iloc functions and perform a describe()

* How many matches have first innings batters lost?


* Get rows 50-200 using iloc functions and perform a describe()

* Find value counts of toss_decision?

## What if I want to select columns by name not index

### Basic indexing

In [60]:
df[ ["winner", "date", "city"] ]

winner        date         city
0    Royal Challengers Bangalore  2010-04-24       Mumbai
1               Delhi Daredevils  2010-03-29        Delhi
2                 Mumbai Indians  2008-05-16       Mumbai
3    Royal Challengers Bangalore  2016-04-22         Pune
4            Chennai Super Kings  2010-04-22       Mumbai
..                           ...         ...          ...
741          Sunrisers Hyderabad  2018-04-12    Hyderabad
742          Chennai Super Kings  2009-05-04  East London
743               Mumbai Indians  2008-05-28    Bangalore
744              Kings XI Punjab  2014-04-28          NaN
745               Mumbai Indians  2010-04-17    Bangalore

[746 rows x 3 columns]

### Using loc

In [61]:
df.loc[1:10, ["city", "date"]]

city        date
1        Delhi  2010-03-29
2       Mumbai  2008-05-16
3         Pune  2016-04-22
4       Mumbai  2010-04-22
5   Chandigarh  2010-04-02
6    Bangalore  2008-05-19
7      Kolkata  2016-05-04
8      Chennai  2011-05-01
9       Mumbai  2011-05-20
10      Mumbai  2014-05-30

# Update data example

## Do it yourself

Use index access to get the columns 

```
['win_by_runs', 'win_by_wickets']
```

Use `loc` to get first 100 rows with columns

```
['win_by_runs', 'win_by_wickets']
```
perform a describe()?

Somethings is different in the output, what is it? and why?


### Quick summary

* Load data into pandas
* `head()`, `tail()` which help you look at data
* `describe()` which gives info about the data frame
* Accessing a single column
* Getting insights from a single columns
* Using iloc and loc functions

---



---




# Break Time

![Break](https://assets.website-files.com/5876c7374691a7d805ce8d19/5f485e12cc612bf82c5ca433_shutterstock_792140977-p-1600.jpeg)